packages/react-storage/src/components/StorageBrowser/Views/Controls/Table.tsx

```diff
@@ -153,57 +155,99 @@ export interface Column<T> {
 
 export interface TableControl<
   T extends StorageBrowserElements = StorageBrowserElements,
-> extends Pick<T, 'TableData' | 'TableRow'> {
+> extends Pick<T, 'TableData' | 'TableRow' | 'TableHeader'> {
   <U>(props: TableControlProps<U>): React.JSX.Element;
 }
 
+export type RenderHeaderItem<T> = (column: Column<T>) => JSX.Element;
+
 export type RenderRowItem<T> = (row: T, index: number) => JSX.Element;
 
+interface TableSortState<T> {
+  direction: SortDirection;
+  selection: keyof T;
+}
+
+type TableSortAction<T> = { selection: keyof T };
+
+export function tableSortReducer<T>(
+  state: TableSortState<T>,
+  action: TableSortAction<T>
+): TableSortState<T> {
+  const { direction, selection: prevSelection } = state;
+  const { selection } = action;
+
+  if (selection === prevSelection) {
+    const newDirection: SortDirection =
+      direction === 'ASCENDING' ? 'DESCENDING' : 'ASCENDING';
+
+    return { direction: newDirection, selection };
+  } else {
+    return { direction: 'ASCENDING', selection };
+  }
+}
+
+export function defaultTableSort<T>(
```

Comment:
Thanks Caleb, these callouts make sense, but I'm confused on how to do it without `selection`. 

Are you suggesting that the sort function we implement should just be the the `compareFn` that we pass into `sort()`? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#comparefn

Don't we need to know what property we're sorting against so that we can reorder the objects?